### PR TITLE
Add table clone and delete from workspace mutation

### PR DESCRIPTION
**Pull request by k9845**
Addresses
 - https://github.com/the-dive/dive-server/issues/23
 
image

 - https://github.com/the-dive/dive-server/issues/8
 
## Changes
 Added following mutations
   - clone table
   - deletefromworkspace
   - add column option properties
  
Mention related users here if any.

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] n+1 queries
- [x] flake8 issues
- [x] `print`
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [x] tests
- [ ] permission checks (tests here too)
- [ ] translations

### CODE REVIEW

File: apps/core/tests/test_mutations.py [hunk 1]
```diff
@@ -160,3 +160,67 @@ def test_update_properties(self):
...
+                }
+            }
+        """
+        resp_data = self.query_check(
+            query,
+            variables={"id": table.id},
+        )
+        content = resp_data["data"]["cloneTable"]
+        assert content["ok"] is True
+        self.assertEqual(Table.objects.count(), old_table_count + 1)
```
Comment (bewakes) [inline apps/core/tests/test_mutations.py:224]: Can also use `assert_object_created` decorator from `base_test.py`.

File: apps/file/utils.py [hunk 2]
```diff
@@ -28,6 +28,8 @@ def process_excel_file(dataset: Dataset):
                 "preview_data": preview_data or {},
                 "has_errored": err is not None,
                 "error": err or None,
+                # Setting this for now, might change later
+                "is_added_to_workspace": True,
```
Comment (bewakes) [inline apps/file/utils.py:32]: This should by default be `False`. I do not understand why it is set `True` here.

**Review by bewakes — CHANGES_REQUESTED** [review]

File: apps/core/tests/test_mutations.py (hunk 1 above)
Comment (k9845) [reply apps/core/tests/test_mutations.py]: Done

File: apps/file/utils.py (hunk 2 above)
Comment (k9845) [reply apps/file/utils.py]: Removed.

File: apps/core/mutations.py [hunk 3]
```diff
@@ -54,10 +55,10 @@ def mutate(root, info, data):
         return CreateDataset(result=dataset, errors=None, ok=True)
 
 
-class UpdateTable(DiveMutationMixin):
+class AddTableToWorkSpace(DiveMutationMixin):
     class Arguments:
         data = TableInputType(required=True)
```
Comment (bewakes) [inline apps/core/mutations.py:60]: Should this not contain just `isAddedToWorkspace`? `TableInputType` will allow for other items to be changed as well. I don't think we want to allow other modifications in this API.

File: apps/core/mutations.py [hunk 4]
```diff
@@ -66,6 +67,97 @@ class Arguments:
...
+    errors = graphene.List(graphene.NonNull(CustomErrorType))
+    ok = graphene.Boolean()
+    result = graphene.Field(TableType)
+
+    @staticmethod
+    def mutate(root, info, id, data):
+        try:
+            instance = Table.objects.get(id=id)
+        except Table.DoesNotExist:
+            return DeleteTableFromWorkspace(
```
Comment (bewakes) [inline apps/core/mutations.py:84]: Should this not be `UpdateTableProperties`? `DeleteTableFromWorkspace` seems like typo for me.

File: apps/core/mutations.py [hunk 5]
```diff
@@ -66,6 +67,97 @@ class Arguments:
...
+        )
+        if errors := mutation_is_not_valid(serializer):
+            return UpdateTableProperties(
+                errors=errors,
+                ok=False,
+            )
+        instance.properties = serializer.data
+        instance.save()
+        # call apply tabel properties
+        apply_properties_to_table()
```
Comment (bewakes) [inline apps/core/mutations.py:103]: We already have a dedicated function for this. You can call the existing function `extract_preview_data` in `utils/extraction.py`.

File: apps/core/mutations.py [hunk 6]
```diff
@@ -66,6 +67,97 @@ class Arguments:
...
+    errors = graphene.List(graphene.NonNull(CustomErrorType))
+    ok = graphene.Boolean()
+    result = graphene.Field(TableType)
+
+    @staticmethod
+    def mutate(root, info, id):
+        try:
+            instance = Table.objects.get(id=id)
+        except Table.DoesNotExist:
+            return DeleteTableFromWorkspace(
```
Comment (bewakes) [inline apps/core/mutations.py:146]: I do not understand why `DeleteTableFromWorkspace` is used here.

File: apps/core/tests/test_mutations.py [hunk 7]
```diff
@@ -118,45 +118,109 @@ def test_add_to_workspace(self):
...
-
         resp_data = self.query_check(
             mutate_query,
-            minput={"properties": valid_props},
+            minput=data,
             variables={"tableId": table.id},
         )
-        content = resp_data["data"]["updateTable"]
+        content = resp_data["data"]["updateTableProperties"]
+        print(content)
```
Comment (bewakes) [inline apps/core/tests/test_mutations.py:224]: Missing print.

File: apps/core/utils.py [hunk 8]
```diff
@@ -0,0 +1,2 @@
+def apply_properties_to_table():
```
Comment (bewakes) [inline apps/core/utils.py:1]: We probably don't need this. You can call the existing function `extract_preview_data` in `utils/extraction.py`.

**Review by bewakes — CHANGES_REQUESTED** [review]

File: apps/core/utils.py (hunk 8 above)
Comment (bewakes) [reply apps/core/utils.py:1]: But never mind. I'll fix this.

File: apps/core/mutations.py (hunk 5 above)
Comment (bewakes) [reply apps/core/mutations.py:103]: Nevermind. I'll fix this.

File: apps/core/mutations.py (hunk 3 above)
Comment (k9845) [reply apps/core/mutations.py]: thanks

File: apps/core/mutations.py (hunk 4 above)
Comment (k9845) [reply apps/core/mutations.py]: Done

File: apps/core/tests/test_mutations.py (hunk 7 above)
Comment (k9845) [reply apps/core/tests/test_mutations.py]: Done.

File: apps/core/mutations.py (hunk 6 above)
Comment (k9845) [reply apps/core/mutations.py]: Done

File: apps/core/mutations.py (hunk 5 above)
Comment (k9845) [reply apps/core/mutations.py]: left with `#TODO`

**Review by bewakes — APPROVED** [review]

**Pull request merged**